### PR TITLE
Updated the LEMP example for docker-compose [skip ci]

### PR DESCRIPTION
**Pull request by chengwang86**
(i) updated docker-compose.yml in"demos/compose/webserving_app/"
(ii) added documentation for the LEMP example  to "vic/doc/application_workflow/"

### CODE REVIEW

File: doc/application_workflow/webserving.md [hunk 1]
```diff
@@ -0,0 +1,137 @@
+# Web-serving on vSphere Integrated Container Engine
```
Comment (anchal-agrawal) [inline doc/application_workflow/webserving.md]: `... Containers Engine`

File: doc/application_workflow/webserving.md [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+We take the Web-serving benchmark from CloudSuite (http://cloudsuite.ch/webserving/) as an example, to demonstrate how customers who are interested in the LEMP implementation of a cloud native web-serving application could use our guidelines to deploy the system on vSphere Integrated Container Engine 0.7.0 with Docker Compose. This demo has three tiers deployed on three containerVMs: an Nginx Web server, a Memcached server, and a Mysql database server. The Web server runs Elgg (a social networking engine) and connects the Memcached server and the database server through network.
```
Comment (chengwang86) [inline doc/application_workflow/webserving.md]: Thanks. I have updated it.

File: doc/application_workflow/webserving.md [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+       - my_net    
+```
+
+### Deploy to Your VCH
+
+Now assume that you already have a VCH deployed by vic-machine. Then go to the folder where you have the above “docker-compose.yml” file and execute the following command to start the Web-serving application:
+```
+$> docker-compose -H VCH_IP:VCH_PORT up –d
+```
+Here VCH_IP and VCH_PORT can be found from the standard output when you use “vic-machine-create” to launch the VCH. Now we are ready to view the Website. Open a browser and navigate to http://web_server:8080. You should be able to see the following page:
```
Comment (anchal-agrawal) [inline doc/application_workflow/webserving.md]: `website`

File: doc/application_workflow/webserving.md [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+![Web serving demo](images/elgg.png)
+
+You can login in as the admin user (username: admin; password: admin1234), or register as a new user with a valid email address (Gmail does not work). You can also create your own contents, invite friends, or chat with others. Enjoy!
```
Comment (anchal-agrawal) [inline doc/application_workflow/webserving.md]: * Why does Gmail not work here?
* `create your own content`

File: doc/application_workflow/webserving.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
+# Web-serving on vSphere Integrated Container Engine
+
+We take the Web-serving benchmark from CloudSuite (http://cloudsuite.ch/webserving/) as an example, to demonstrate how customers who are interested in the LEMP implementation of a cloud native web-serving application could deploy it on vSphere Integrated Container Engine 0.7.0 with Docker Compose. This demo has three tiers deployed on three containerVMs: an Nginx Web server, a Memcached server, and a MySQL database server. The Web server runs Elgg (a social networking engine) and connects the Memcached server and the database server through the network.
+
+## Workflow
+
+### Build docker image for the Web server (on regular docker)
+
+In the original the Web-server docker image from Cloudsuite, the functionality of “email verification for new user registration” is not enabled, which makes it less realistic and practical. Therefore, we need make some modifications and re-build the docker image for the Web server. **You can also skip this section and proceed to "Compose File for vSphere Integrated Containers Engine" if you do not want to build your own image**.
```
Comment (anchal-agrawal) [inline doc/application_workflow/webserving.md]: * Did you mean `In the original Web-server docker image`?
* You can link directly to the `Compose File for vSphere Integrated Containers Engine` section using markdown with `#compose-file-for-vsphere-integrated-containers-engine`

File: doc/application_workflow/webserving.md [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+cat /etc/hosts
+service sendmail stop
+service sendmail start
+service php5-fpm restart
+service nginx restart
+```
+Step IV: (In this example, we will deploy the image to the docker hub.)
+-	Build the image: 
+```
+$> docker build  -t repo/directory:tag.
```
Comment (anchal-agrawal) [inline doc/application_workflow/webserving.md]: Is the `.` after `tag` intentional?

File: doc/application_workflow/webserving.md [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+-	Login to your registry: 
+```
+$> docker login (input your credentials when needed)
+```
+-	upload your image: 
+```
+$> docker push repo/directory:tag
+```
+
+### Build docker image for the Mysql server (on regular docker)
```
Comment (anchal-agrawal) [inline doc/application_workflow/webserving.md]: `MySQL`

File: doc/application_workflow/webserving.md [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+THe Web server reads from the database to obtain the "site_url" to generate links for the Web pages. However, the original database image from cloudsuite (cloudsuite/web-serving:db_server) is configured with pre-defined environment variable "env web_host web_server" in the Dockerfile, which is then used to pre-dump the ELGG database with static "site_url" of "http://web_server:8080" in the image; thus the url is not accessible when deployed on the VCH. Therefore, we need to modify the database image: allow the database to be populated at the start of the containerVM using the environment variable ${web_host}; then you can push the new image to your own registry. **You can also skip this section and proceed to "Compose File for vSphere Integrated Containers Engine" if you do not want to build your own image**.
```
Comment (anchal-agrawal) [inline doc/application_workflow/webserving.md]: ${web_host} -> \`web_host\`

File: doc/application_workflow/webserving.md [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+# Need bash -c for redirection
+bash -c "mysql -uroot -p$root_password ELGG_DB < /elgg_db.dump"
+
+mysql -uroot -p$root_password -e "GRANT ALL PRIVILEGES ON *.* TO 'root'@'%' IDENTIFIED BY '$root_password' WITH GRANT OPTION; FLUSH PRIVILEGES;"
+
+service mysql stop 
+/usr/sbin/mysqld
+```
+
+Step IV: The same with the Step IV when creating the docker image for the Web server.
```
Comment (anchal-agrawal) [inline doc/application_workflow/webserving.md]: `Same as Step IV...`

File: doc/application_workflow/webserving.md [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+   memcache_server:
+     image: cloudsuite/web-serving:memcached_server
+     container_name: memcache_server
+     networks:
+       - my_net    
+```
+
+### Deploy to Your VCH
+
+Now assume that you already have a VCH deployed by vic-machine. Then go to the folder where you have the above “docker-compose.yml” file and execute the following command to start the Web-serving application:
```
Comment (anchal-agrawal) [inline doc/application_workflow/webserving.md]: `Once you already have a VCH deployed by vic-machine, go to...`

File: doc/application_workflow/webserving.md (hunk 8 above)
Comment (anchal-agrawal) [inline doc/application_workflow/webserving.md]: `THe` -> The

File: doc/application_workflow/webserving.md (hunk 4 above)
Comment (chengwang86) [inline doc/application_workflow/webserving.md]: By default, the notification emails are sent to the users from "admin@elgg-community.com" (a fake email address) which might have been blocked by gmail (I guess). But it works well if I use my corporate email address to register a new user.

File: demos/compose/webserving_app/docker-compose.yml [hunk 11]
```diff
@@ -3,26 +3,31 @@ version: '2'
...
   my_net:
     driver: bridge
-services:
 
+services:
   web_server:
-  image: wangcheng86/wangcheng86:web_mail_enabled
-  #image: your_own_image_built_before
-    container_name: web_server    
+    #image: your_own_web_image
```
Comment (sflxn) [inline demos/compose/webserving_app/docker-compose.yml]: Remove unused code

File: demos/compose/webserving_app/docker-compose.yml [hunk 12]
```diff
@@ -3,26 +3,31 @@ version: '2'
...
     networks:
        - my_net
     ports:
       - "8080:8080"
       - "25:25"
-      
+
   mysql_server:
-    image: cloudsuite/web-serving:db_server
+    #image: your_own_db_image
```
Comment (sflxn) [inline demos/compose/webserving_app/docker-compose.yml]: Same.  Don't leave commented code behind.

File: demos/compose/webserving_app/docker-compose.yml (hunk 11 above)
Comment (chengwang86) [inline demos/compose/webserving_app/docker-compose.yml]: Shouldn't we tell the customers where to place their own images?

File: doc/application_workflow/webserving.md [hunk 13]
```diff
@@ -0,0 +1,137 @@
+# Web-serving on vSphere Integrated Container Engine
+
+We take the Web-serving benchmark from CloudSuite (http://cloudsuite.ch/webserving/) as an example, to demonstrate how customers who are interested in the LEMP implementation of a cloud native web-serving application could deploy it on vSphere Integrated Container Engine 0.7.0 with Docker Compose. This demo has three tiers deployed on three containerVMs: an Nginx Web server, a Memcached server, and a MySQL database server. The Web server runs Elgg (a social networking engine) and connects the Memcached server and the database server through the network.
```
Comment (chengwang86) [inline doc/application_workflow/webserving.md]: In Loc's voting app example, he still has the version number. So I think I should also put it here.

**Comment by anchal-agrawal** [comment]
Please rename `demos/compose/webserving_app` to `demos/compose/webserving-app` to be consistent with the other folder's name (`voting-app`).

**Comment by anchal-agrawal** [comment]
Thanks @chengwang86! LGTM

File: doc/application_workflow/webserving.md [hunk 14]
```diff
@@ -1,4 +1,4 @@
-# Web-serving on vSphere Integrated Container Engine
+# Web-serving on vSphere Integrated Containers Engine
 
 We take the Web-serving benchmark from CloudSuite (http://cloudsuite.ch/webserving/) as an example, to demonstrate how customers who are interested in the LEMP implementation of a cloud native web-serving application could deploy it on vSphere Integrated Containers Engine 0.7.0 with Docker Compose. This demo has three tiers deployed on three containerVMs: an Nginx Web server, a Memcached server, and a MySQL database server. The Web server runs Elgg (a social networking engine) and connects the Memcached server and the database server through the network.
```
Comment (sflxn) [inline doc/application_workflow/webserving.md]:
```
to demonstrate how customers who are interested in the LEMP implementation of a web-serving application can deploy it on vSphere Integrated Containers Engine 0.7.0 using Docker Compose.
```
I would reword it to the above.

File: doc/application_workflow/webserving.md [hunk 15]
```diff
@@ -0,0 +1,137 @@
+# Web-serving on vSphere Integrated Containers Engine
+
+We take the Web-serving benchmark from CloudSuite (http://cloudsuite.ch/webserving/) as an example, to demonstrate how customers who are interested in the LEMP implementation of a cloud native web-serving application could deploy it on vSphere Integrated Containers Engine 0.7.0 with Docker Compose. This demo has three tiers deployed on three containerVMs: an Nginx Web server, a Memcached server, and a MySQL database server. The Web server runs Elgg (a social networking engine) and connects the Memcached server and the database server through the network.
+
+## Workflow
+
+### Build docker image for the Web server (on regular docker)
+
+In the original Web-server docker image from Cloudsuite, the functionality of “email verification for new user registration” is not enabled, which makes it less realistic and practical. Therefore, we need make some modifications and re-build the docker image for the Web server. **You can also skip this section and proceed to "[Compose File for vSphere Integrated Containers Engine](#compose-file-for-vsphere-integrated-containers-engine)" if you do not want to build your own image**.
```
Comment (sflxn) [inline doc/application_workflow/webserving.md]:
```
In the original Web-server docker image from Cloudsuite, the functionality of “email verification for new user registration” is not enabled, which makes it less realistic and practical. Therefore, we need make some modifications and re-build the docker image for the Web server. 
```

How about,

```
Note, in the original web-server docker image from Cloudesuite, the email verification for new user is not enabled.
```

File: doc/application_workflow/webserving.md [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+$> docker login (input your credentials when needed)
+```
+-	upload your image: 
+```
+$> docker push repo/directory:tag
+```
+
+### Build docker image for the MySQL server (on regular docker)
+
+The Web server reads from the database to obtain the "site_url" to generate links for the Web pages. However, the original database image from cloudsuite (cloudsuite/web-serving:db_server) is configured with pre-defined environment variable "env web_host web_server" in the Dockerfile, which is then used to pre-dump the ELGG database with static "site_url" of "http://web_server:8080" in the image; thus the url is not accessible when deployed on the VCH. Therefore, we need to modify the database image: allow the database to be populated at the start of the containerVM using the environment variable "web_host"; then you can push the new image to your own registry. **You can also skip this section and proceed to "[Compose File for vSphere Integrated Containers Engine](#compose-file-for-vsphere-integrated-containers-engine)" if you do not want to build your own image**.
```
Comment (sflxn) [inline doc/application_workflow/webserving.md]: This is difficult to read.  Let me take a crack at this.

```
This example application uses the database to store the address of the web server.  The original Dockerfile from Cloudsuite populates this with "http://web_server:8080", which is not usable in production.  Using the suggestions we provided earlier, we modify the execute.sh script to replace the "web_server" text to the actual IP of our VCH.  This script is executed when this database container is executed.  In the Docker Compose file, we specify the IP address of our target VCH.  You will see that in the modified compose yml file below.  This example illustrates using a script and command line arguments to pass config to a containerized app.  Another option is to pass config in via environment variables and have the script use those values to modify internal config in the running container.

Below, we will modify the Dockerfile and script. **You can also skip this section and proceed to "[Compose File for vSphere Integrated Containers Engine](#compose-file-for-vsphere-integrated-containers-engine)" if you do not want to build your own image**.
```

File: doc/application_workflow/webserving.md [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+    image: victest/web_elgg
+    container_name: web_server
+    networks:
+       - my_net
+    ports:
+      - "8080:8080"
+      - "25:25"
+
+  mysql_server:
+    #image: your_own_db_image
```
Comment (sflxn) [inline doc/application_workflow/webserving.md]: remove that line

File: doc/application_workflow/webserving.md [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+```
+version: '2'
+
+networks:
+  my_net:
+    driver: bridge
+
+services:
+  web_server:
+    #image: your_own_web_image
```
Comment (sflxn) [inline doc/application_workflow/webserving.md]: remove that line

File: doc/application_workflow/webserving.md [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+       - my_net    
+```
+
+### Deploy to Your VCH
+
+Once you already have a VCH deployed by vic-machine, go to the folder where you have the above “docker-compose.yml” file and execute the following command to start the Web-serving application:
+```
+$> docker-compose -H VCH_IP:VCH_PORT up –d
+```
+Here VCH_IP and VCH_PORT can be found from the standard output when you use “vic-machine-create” to launch the VCH. Now we are ready to view the website. Open a browser and navigate to http://web_server:8080. You should be able to see the following page:
```
Comment (sflxn) [inline doc/application_workflow/webserving.md]: "vic-machine create", no hyphen.

navigate to http://VCH_IP:8080.  We make sure to use the IP address of the VCH as the IP of our webserver here.

**Comment by sflxn** [comment]
lgtm.  Update your branch and rebase.

**Comment by chengwang86** [comment]
ping
